services/system/Transact/src/RTransact.cpp

```diff
@@ -177,129 +179,327 @@ namespace
       }
    };
 
-}  // namespace
+   using ClientFilter = std::function<bool(const TraceClientInfo&)>;
+   bool noFilter(const TraceClientInfo&)
+   {
+      return true;
+   }
 
-void RTransact::onTrx(const Checksum256& id, psio::view<const TransactionTrace> trace)
-{
-   check(getSender() == AccountNumber{}, "Wrong sender");
-   printf("trace size: %zu\n", find_view_span(trace).size());
+   // Returns a tuple containing:
+   // 1. Vector of sockets waiting for a JSON response
+   // 2. Vector of sockets waiting for a binary response
+   //
+   // Removes any clients from TraceClientTable for whom replies have been claimed.
+   auto claimClientReply(const psibase::Checksum256& id, const ClientFilter& clientFilter)
+       -> std::tuple<std::vector<std::int32_t>, std::vector<std::int32_t>>
+   {
+      {
+         auto                      clients = RTransact::Subjective{}.open<TraceClientTable>();
+         std::vector<std::int32_t> json_clients, bin_clients;
 
-   TransactionTraceRef pruned = PruneTrace{true}(trace);
+         auto socketClaimed = [&](const auto& client)
+         {
+            return std::ranges::contains(json_clients, client.socket) ||
+                   std::ranges::contains(bin_clients, client.socket);
+         };
 
-   auto                          clients = Subjective{}.open<TraceClientTable>();
-   std::optional<TraceClientRow> row;
-   bool                          json;
-   bool                          bin;
-   PSIBASE_SUBJECTIVE_TX
+         PSIBASE_SUBJECTIVE_TX
+         {
+            if (auto row_opt = clients.get(id))
+            {
+               auto& row = *row_opt;
+               if (std::ranges::none_of(row.clients, clientFilter))
+                  abortMessage("No clients matching filter to claim a reply for this tx");
```

Comment:
Can the caller actually guarantee this? `sendReply` checks whether there were any clients *after* calling `claimClientReply`.

Reply:
What about this:
- send replies
- switch forks in a way the rolls back commitNum (Note: it isn't possible to fork out the block that is committed, but it *is* possible to switch to a fork that has a lower `commitNum` in the HEAD block)
- A new client pushes the same transaction with wait_for=applied

Now, we have both a successful trace and a client, but the conditions that caused the trace to be saved were based on the previous client, so the invariant that we're expecting doesn't hold.

Reply:
> I'm realizing that there are problems with pushing a transaction that has already been applied.

If you're not going to look into it right away, mind adding a github issue to track this?

Reply:
In my scenario, the transaction is *not* executed again.

Reply:
We're only forking back to a point before it was declared committed, not before it was executed.